README
Business-SiteCatalyst

This module allows you to interact with Adobe (formerly Omniture) SiteCatalyst,
a web analytics service. It encapsulates all the communications with the API 
provided by Adobe SiteCatalyst to offer a Perl interface for managing reports,
pulling company-specific SiteCatalyst data (ex: token usage), uploading SAINT 
data (feature not implemented yet), etc.

Please note that you will need to have purchased the Adobe SiteCatalyst product,
and have web services enabled within your account first in order to obtain a web
services shared secret, as well as agree with the Terms and Conditions for using 
the API.

NOTE: the 'api_subdomain' option/config variable is utilized for the api url.
To determine your specific API URL/Endpoint, please visit
https://developer.omniture.com/en_US/get-started/api-explorer
Most users won't need to set this variable unless the default causes errors.

API URL: 'https://' . $api_subdomain . '.omniture.com/admin/1.3/rest/?'


INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Business::SiteCatalyst

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Business-SiteCatalyst

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Business-SiteCatalyst

    CPAN Ratings
        http://cpanratings.perl.org/d/Business-SiteCatalyst

    Search CPAN
        http://search.cpan.org/dist/Business-SiteCatalyst/


LICENSE AND COPYRIGHT

Copyright (C) 2012 Jennifer Pinkham

This program is free software; you can redistribute it and/or modify it
under the terms of the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Business-SiteCatalyst

        10/18/2017
        Modified README to add a note that this module is up for adoption.

v1.2.2  03/24/2013
        Refactored Business::SiteCatalyst to use POSIX instead of DateTime,
          to remove a dependancy on Params::Util (which prevented running
          Devel::Cover against the code).
        Updated documentation.


v1.2.1  11/11/2012
        Updated all subs in Business::SiteCatalyst::Company to croak if
            there is no response from the API method call.
        Updated repository path. Project moved from SourceForge to GitHub.
        Updated documentation.


v1.2.0  11/10/2012
        Added new Business::SiteCatalyst::Company module, for the 'Company'
            API within SiteCatalyst. Features include: getting token usage,
            determine tracking server, report suite list, and more.
        Refactored to utilize an optional config variable 'api_subdomain' to
            solve problem of some companies connecting to api2.omniture.com
            (vs default of api.omniture.com). Thanks to Gianluca Casati for
            pointing this out.
        Updated Business::SiteCatalyst::Report::cancel() so it actually works.
        Updated documentation.


v1.0.1  04/22/2012
        Updated documentation.


v1.0.0  04/20/2012
        First release.


Build.PL
use 5.006;
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
	module_name          => 'Business::SiteCatalyst',
	license              => 'artistic',
	dist_author          => q|Jennifer Pinkham <jpinkham@cpan.org>|,
	dist_version_from    => 'lib/Business/SiteCatalyst.pm',
	build_requires       =>
	{
		'Carp'                 => 0,
		'Data::Dumper'         => 0,
		'Data::Validate::Type' => "1.1.0",
		'DateTime'             => 0,
		'Digest::MD5'          => 0,
		'Digest::SHA1'         => 0,
		'HTTP::Request'        => 0,
		'JSON'                 => 0,
		'LWP::UserAgent'       => 0,
		'MIME::Base64'         => 0,
		'Test::More'           => 0.94,
		'Try::Tiny'            => 0,
	},
	requires             =>
	{
		'perl' => 5.006,
	},
	add_to_cleanup       => [ 'Business-SiteCatalyst-*' ],
	create_makefile_pl   => 'traditional',
	recursive_test_files => 1,
	configure_requires   =>
	{
		'Module::Build'  => 0,
	},
	meta_merge           =>
	{
		resources        =>
		{
			repository  => 'https://github.com/jpinkham/business-sitecatalyst.git',
			homepage    => 'https://metacpan.org/release/Business-SiteCatalyst',
			bugtracker  => 'https://github.com/jpinkham/business-sitecatalyst/issues',
		},
	},
);

$builder->create_build_script();


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.4003
require 5.006;
use ExtUtils::MakeMaker;
WriteMakefile
(
  'NAME' => 'Business::SiteCatalyst',
  'VERSION_FROM' => 'lib/Business/SiteCatalyst.pm',
  'PREREQ_PM' => {
                   'Carp' => 0,
                   'Data::Dumper' => 0,
                   'Data::Validate::Type' => '1.001000',
                   'Digest::MD5' => 0,
                   'Digest::SHA1' => 0,
                   'HTTP::Request' => 0,
                   'JSON' => 0,
                   'LWP::UserAgent' => 0,
                   'MIME::Base64' => 0,
                   'Test::More' => '0.94',
                   'Try::Tiny' => 0
                 },
  'INSTALLDIRS' => 'site',
  'EXE_FILES' => [],
  'PL_FILES' => {},
  'test' => {
              'TESTS' => 't/*.t t/Report/*.t t/Company/*.t'
            }
)
;


lib/Business/SiteCatalyst/Report.pm
package Business::SiteCatalyst::Report;

use strict;
use warnings;

use Carp;
use Data::Dumper;
use Data::Validate::Type;


=head1 NAME

Business::SiteCatalyst::Report - Interface to Adobe Omniture SiteCatalyst's REST API 'Report' module.


=head1 VERSION

Version 1.2.2

=cut

our $VERSION = '1.4.0';


=head1 SYNOPSIS

This module allows you to interact with Adobe (formerly Omniture) SiteCatalyst,
a web analytics service. It encapsulates all the communications with the API 
provided by Adobe SiteCatalyst to offer a Perl interface for managing reports,
pulling company-specific SiteCatalyst data (ex: token usage), uploading SAINT 
data (feature not implemented yet), etc.

Please note that you will need to have purchased the Adobe SiteCatalyst product,
and have web services enabled within your account first in order to obtain a web
services shared secret, as well as agree with the Terms and Conditions for using
the API.

	use Business::SiteCatalyst;
	
	# Create an object to communicate with Adobe SiteCatalyst
	my $site_catalyst = Business::SiteCatalyst->new(
		username        => 'dummyusername',
		shared_secret   => 'dummysecret',
		api_subdomain   => 'api|api2|api3|api4|api5', #optional; default value='api'
	);
	
	my $report = $site_catalyst->instantiate_report(
		type            => 'report type',
		report_suite_id => 'report suite id',
	);
	
	# See SiteCatalyst API Explorer at
	# https://developer.omniture.com/en_US/get-started/api-explorer
	# for Report.Queue[Trended|Ranked|Overtime|Pathing|Fallout|Summary|Real-Time] documentation
	
	$report->queue(
		%report_arguments, #report-dependant
	);
	
	my $results;
	for ( my $tries = 0; $tries < 20; tries++ )
	{
		if ( $report->is_ready() )
		{
			$results = $report->retrieve();
			last;
		}
		else
		{
			sleep 5;
		}
	}
	
	unless ( defined $results )
	{
		$cancel_success = $report->cancel();
	}

=head1 METHODS

=head2 new()

Create a new Business::SiteCatalyst::Report object, which
will allow retrieval of SiteCatalyst reports.

NOTE: This should not be called directly. Instead, use C<Business::SiteCatalyst->instantiate_report()>.

	my $report = Business::SiteCatalyst::Report->new(
		$site_catalyst,
		report_suite_id => 'report suite id',
	);

Parameters:

=over 4

=item * report_suite_id

The Report Suite ID you want to pull data from.

=back

=cut

sub new
{
	my ( $class, $site_catalyst, %args ) = @_;
	
	# Check for mandatory parameters
	Data::Validate::Type::is_instance( $site_catalyst, class => 'Business::SiteCatalyst')
		|| croak "First argument must be a Business::SiteCatalyst object";

	my $mode = defined $args{'report_id'} ? "existing" : "new";
		
	my %required_arguments = (
		'report_id'       => 'existing',
		'report_suite_id' => 'new',
	);
		
	foreach my $arg ( keys %required_arguments )
	{
		if ( $required_arguments{ $arg } eq $mode )
		{
			croak "Argument '$arg' is needed to create the Business::SiteCatalyst::Report object"
				if !defined( $args{$arg} ) || ( $args{$arg} eq '' );
		}
		else
		{
			croak "Argument '$arg' is not valid when loading '$mode' report"
				if defined( $args{$arg} );
		}
	}

	# Create the object
	my $self = bless(
		{
			site_catalyst   => $site_catalyst,
			report_suite_id => $args{'report_suite_id'},
			id              => $args{'report_id'},
		},
		$class,
	);
	
	return $self;
}


=head2 queue()

Queue a Business::SiteCatalyst report.

	$report->queue( %report_arguments );

	# Example: Top 5 referrers report
	$report->queue(
		dateFrom      => "2012-04-01",
		dateTo        => "2012-04-15",
		metrics       => [{"id" => "instances"}],
		elements      => [{"id" => "referrer","top" => "5"}]
	);

=cut

sub queue
{
	my ( $self, %args ) = @_;
	
	$self->{report_response} = undef;
	
	my $site_catalyst = $self->get_site_catalyst();
	my $verbose = $site_catalyst->verbose();

	my $response = $site_catalyst->send_request(
		method    => 'Report.Queue',
		data      =>
		{
			reportDescription =>
				{
					reportSuiteID => $self->{'report_suite_id'},
					%args,
				}
		},
		test_mode => defined $args{'test_mode'} ? $args{'test_mode'} : 0,
	);
	
	if ( !defined($response) )
	{
		croak "Fatal error. No response.";
	}
	elsif ( !defined($response->{'reportID'}) )
	{
		carp "Full response: " . Dumper($response) if $verbose;
		croak "Fatal error. Missing reportID in response.";
	}

	# Store report id; we'll need it to check the status
	$self->{'id'} = $response->{'reportID'};
	
	return $response;
}


=head2 is_ready()

Check if a queued report is completed yet.

	my $boolean = $report->is_ready();

=cut

sub is_ready
{
	my ( $self, %args ) = @_;
	
	return $self->get( %args );
	
}


=head2 retrieve()

Return report results from previously fetched by calling
is_ready().

	my $results = $report->retrieve();

=cut

sub retrieve
{
	my ( $self, %args ) = @_;
	
	if ( ! $self->{report_response} ) {
		croak "Please call is_ready() before attempting retrieval. Report is not done.";
	}
	return $self->{report_response};
}

=head2 get()

Calls Report.Get to get report results from Adobe SiteCatalyst,
OR the 'error': report_not_valid. If the report is ready, it is stored
in memory so retrieve() can return it.

	my $results = $report->get();

=cut

sub get
{
	my ( $self, %args ) = @_;
		
	my $site_catalyst = $self->get_site_catalyst();
	my $verbose = $site_catalyst->verbose();
	
	my $response = $site_catalyst->send_request(
		method    => 'Report.Get',
		data      =>
		{
			reportID => $self->get_id(),
		},
		test_mode => defined $args{'test_mode'} ? $args{'test_mode'} : 0,
		content_on_error => 1,       # adobe send a 400 error if the report is not ready
	);
	
	if (
		  !defined($response)
	 || ( !defined($response->{'runSeconds'}) && !defined($response->{'error'}) )
	) {
		croak "Fatal error. No response, or missing [ runSeconds and error ] in response";
	}
	elsif ( $response->{'status'} eq 'error' || $response->{'status'} eq 'failed' ) # are these still valid in 1.4?
	{
		carp "Full response: " . Dumper($response) if $verbose;
		croak "Something went wrong with this report!";
	}
	elsif ( defined($response->{'error'}) && $response->{'error'} eq 'report_not_ready' )
	{
		carp "Report not ready";
		return 0
	}
	elsif ( !defined($response->{'report'}) )
	{
		croak "Unhandled response in get():\n" . to_json( $response, { pretty => 1 } );
	}
	
	$self->{report_response} = $response->{'report'};
	
	return 1;
}

=head2 cancel()

Cancel previously submitted report request, and removes it from processing queue.
Returns 1 if successful, otherwise 0.

	my $cancel_success = $report->cancel();

=cut

sub cancel
{
	my ( $self, %args ) = @_;
	
	my $site_catalyst = $self->get_site_catalyst();
	my $verbose = $site_catalyst->verbose();
	
	my $response;
	if ( defined $args{'test_mode'} && $args{'test_mode'} == 1 )
	{
		$response = undef;
	}
	else
	{
		$response = $site_catalyst->send_request(
			method    => 'Report.CancelReport',
			data      =>
			{
				reportID => $self->get_id(),
			},
		);
	}
	
	if ( !defined($response) )
	{
		croak "Fatal error. No response.";
	}
	
	return $response;
}

=head2 get_elements()

Returns a list of available elements for a given report type,
optionally in conjunction with a given list of elements and/or
metrics.

	my $available_elements = $report->get_elements(
            reportType          => 'ranked'                 # optional
          , existingElements    => [ list, of, elements ]   # optional
          , existingMetrics     => [ list, of, metrics ]    # optional
	);

=cut

sub get_elements
{
	my ( $self, %args ) = @_;
	
	my $site_catalyst = $self->get_site_catalyst();
	my $verbose = $site_catalyst->verbose();
	
	my $response;
	if ( defined $args{'test_mode'} && $args{'test_mode'} == 1 )
	{
		$response = undef;
	}
	else
	{
		$response = $site_catalyst->send_request(
			method    => 'Report.GetElements',
			data      =>
			{
				reportSuiteID => $self->{'report_suite_id'},
				%args,
			},
		);
	}
	
	if ( !defined($response) )
	{
		croak "Fatal error. No response.";
	}
	
	return $response;
}

=head2 get_metrics()

Returns a list of available metrics for a given report type,
optionally in conjunction with a given list of elements and/or
metrics.

	my $available_elements = $report->get_elements(
            reportType          => 'ranked'                 # optional
          , existingElements    => [ list, of, elements ]   # optional
          , existingMetrics     => [ list, of, metrics ]    # optional
	);

=cut

sub get_metrics
{
	my ( $self, %args ) = @_;
	
	my $site_catalyst = $self->get_site_catalyst();
	my $verbose = $site_catalyst->verbose();
	
	my $response;
	if ( defined $args{'test_mode'} && $args{'test_mode'} == 1 )
	{
		$response = undef;
	}
	else
	{
		$response = $site_catalyst->send_request(
			method    => 'Report.GetMetrics',
			data      =>
			{
				reportSuiteID => $self->{'report_suite_id'},
				%args,
			},
		);
	}
	
	if ( !defined($response) )
	{
		croak "Fatal error. No response.";
	}
	
	return $response;
}

=head2 get_site_catalyst()

Get Business::SiteCatalyst object used when creating the current object.

	my $site_catalyst = $report->get_site_catalyst();

=cut

sub get_site_catalyst
{
	my ( $self ) = @_;
	
	return $self->{'site_catalyst'};
}


=head2 get_id()

Get the report ID returned by Adobe SiteCatalyst when we queued the report.

	my $report_id = $report->get_id();

=cut

sub get_id
{
	my ( $self ) = @_;
	
	return $self->{'id'};
}


=head1 AUTHOR

Jennifer Pinkham, C<< <jpinkham at cpan.org> >>.


=head1 BUGS

Please report any bugs or feature requests to C<bug-Business-SiteCatalyst at rt.cpan.org>,
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Business-SiteCatalyst>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

	perldoc Business::SiteCatalyst::Report


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Business-SiteCatalyst>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Business-SiteCatalyst>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Business-SiteCatalyst>

=item * Search CPAN

L<http://search.cpan.org/dist/Business-SiteCatalyst/>

=back


=head1 ACKNOWLEDGEMENTS

Thanks to ThinkGeek (L<http://www.thinkgeek.com/>) and its corporate overlords
at Geeknet (L<http://www.geek.net/>), for footing the bill while I write code for them!
Special thanks for technical help from fellow ThinkGeek CPAN author Guillaume Aubert L<http://search.cpan.org/~aubertg/>


=head1 COPYRIGHT & LICENSE

Copyright 2017 Jennifer Pinkham.

This program is free software; you can redistribute it and/or modify it
under the terms of the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;


lib/Business/SiteCatalyst.pm
package Business::SiteCatalyst;

use strict;
use warnings;

use Data::Dumper;
use Carp;
use LWP::UserAgent qw();
use HTTP::Request qw();
use JSON qw();
use Digest::MD5 qw();
use POSIX qw();
use Digest::SHA1 qw();
use MIME::Base64 qw();

use Business::SiteCatalyst::Company;
use Business::SiteCatalyst::Report;


# Some API methods return strings or numbers rather than valid JSON.
# This list is used to return the raw response to the caller instead of decoding output as JSON
my %METHODS_RETURNING_INVALID_JSON = 
(
	'Company.GetEndpoint'   => 1,  # Returns string
	'Company.GetTokenCount' => 1,  # Returns number
	'Report.CancelReport'   => 1,  # Returns number
);


=head1 NAME

Business::SiteCatalyst - Interface to Adobe Omniture SiteCatalyst's REST API.


=head1 VERSION

Version 1.4.0

=cut

our $VERSION = '1.4.0';


=head1 SYNOPSIS

This module allows you to interact with Adobe (formerly Omniture) SiteCatalyst,
a web analytics service. It encapsulates all the communications with the API 
provided by Adobe SiteCatalyst to offer a Perl interface for managing reports,
pulling company-specific SiteCatalyst data (ex: token usage), uploading SAINT 
data (feature not implemented yet), etc.

Please note that you will need to have purchased the Adobe SiteCatalyst product,
and have web services enabled within your account first in order to obtain a web
services shared secret, as well as agree with the Terms and Conditions for using 
the API.

NOTE: the 'api_subdomain' option/config variable is utilized for the api url.
To determine your specific API URL/Endpoint, please visit
https://developer.omniture.com/en_US/get-started/api-explorer
Most users won't need to set this variable unless the default causes errors.

API URL: 'https://' . $api_subdomain . '.omniture.com/admin/1.3/rest/?'


	use Business::SiteCatalyst;
	
	# Create an object to communicate with Adobe SiteCatalyst
	my $site_catalyst = Business::SiteCatalyst->new(
		username        => 'dummyusername',
		shared_secret   => 'dummysecret',
		api_subdomain   => 'api|api2|api3|api4|api5', #optional; default value='api'
	);

The API subdomain indicates the data centre that services the request. Currently
available subdomains are:

* api - San Jose
* api2 - Dallas
* api3 - London
* api4 - Singapore
* api5 - Pacific Northwest

According to Adobe, "Company.GetEndpoint to periodically refresh the endpoint
programmatically, in case the URL changes"

=head1 METHODS

=head2 new()

Create a new Adobe SiteCatalyst object that will be used as the interface with
Adobe SiteCatalyst's API

	use Business::SiteCatalyst;
	
	# Create an object to communicate with Adobe SiteCatalyst
	my $site_catalyst = Business::SiteCatalyst->new(
		username        => 'dummyusername',
		shared_secret   => 'dummysecret',
		api_subdomain   => 'api|api2|api3|api4|api5', #optional; default value='api'
	);

Creates a new object to communicate with Adobe SiteCatalyst.

'username' and 'shared_secret' are mandatory.
The 'verbose' parameter is optional and defaults to not verbose.

=cut

sub new
{
	my ( $class, %args ) = @_;
	
	# Check for mandatory parameters
	foreach my $arg ( qw( username shared_secret ) )
	{
		croak "Argument '$arg' is needed to create the Business::SiteCatalyst object"
			if !defined( $args{$arg} ) || ( $args{$arg} eq '' );
	}
	
	#Defaults.
	
	# NOTE - some users connect to api2.omniture.com, so the subdomain portion of the host is configurable
	# in the 'api_subdomain' config variable in SiteCatalystConfig.pm
	my $webservice_url = 'https://' .
		( defined $args{'api_subdomain'} ? $args{'api_subdomain'} : 'api' ) .
		'.omniture.com/admin/1.4/rest/?method=';
	
	# Create the object
	my $self = bless(
		{
			username        => $args{'username'},
			shared_secret   => $args{'shared_secret'},
			webservice_url  => $webservice_url,
		},
		$class,
	);
	
	$self->verbose( $args{'verbose'} );
	
	return $self;
}


=head2 instantiate_report()

Create a new Business::SiteCatalyst::Report object, which
will allow retrieval of SiteCatalyst reports.

	# Create a new report
	my $report = $site_catalyst->instantiate_report(
		type            => 'report type',
		report_suite_id => 'report suite id',
	);

	# Act on an existing report
	my $report = $site_catalyst->instantiate_report(
		report_id       => 'report id',
	);

	
Parameters:

=over 4

=item * type

The type of the report to instantiate. 
Acceptable values are 'Overtime', 'Ranked', and 'Trended'.

=item * report_suite_id

The Report Suite ID you want to pull data from.

=item * report_id

The id of the existing report you want to check status of, retrieve results for,
or cancel processing.

=back

=cut

sub instantiate_report
{
	my ( $self, %args ) = @_;
	
	return Business::SiteCatalyst::Report->new( $self, %args );
}



=head2 instantiate_company()

Create a new Business::SiteCatalyst::Company object, which
will allow retrieval of company-specific SiteCatalyst data.

	my $company = $site_catalyst->instantiate_company();

	
Parameters: none

=cut

sub instantiate_company
{
	my ( $self, %args ) = @_;
	
	return Business::SiteCatalyst::Company->new( $self, %args );
}


=head1 INTERNAL METHODS

=head2 send_request()

Internal, formats the JSON call with the arguments provided and checks the
reply.

	my ( $error, $response_data ) = $site_catalyst->send_request(
		method    => $method,
		data      => $data,
		test_mode => 0|1, #optional
	);

=cut

sub send_request
{
	my ( $self, %args ) = @_;
	
	my $verbose = $self->verbose();
	my $url = $self->{'webservice_url'} .  $args{'method'};
	my $test_mode = defined $args{'test_mode'} ? $args{'test_mode'} : 0;
	
	# Check for mandatory parameters
	foreach my $arg ( qw( method data ) )
	{
		croak "Argument '$arg' is needed to send a request with the Business::SiteCatalyst object"
			if !defined( $args{$arg} ) || ( $args{$arg} eq '' );
	}
	
	my $json_in = JSON::encode_json( $args{'data'} );
	carp "Sending JSON request >" . ( defined( $json_in ) ? $json_in : '' ) . "<"
		if $verbose;
	
	# Authentication information for header.
	my $username = $self->{'username'};
	my $nonce = Digest::MD5::md5_hex( rand() * time() );
	chomp($nonce);
	
	my $created = POSIX::strftime("%Y-%m-%dT%H:%M:%S", gmtime()) . 'Z';
	my $password_digest = MIME::Base64::encode_base64(
		Digest::SHA1::sha1_hex( $nonce . $created . $self->{'shared_secret'} ) 
	);
	chomp($password_digest);
	
	my $request = HTTP::Request->new(POST => $url);
	carp "POSTing request to URL >" . ( defined( $url ) ? $url : '' ) . "<"
		if $verbose;
	my $auth_header = qq|UsernameToken Username="$username", PasswordDigest="$password_digest", Nonce="$nonce", Created="$created"|;
	carp "Auth header: >$auth_header<" if $verbose;
	
	$request->header('X-WSSE', $auth_header);
	
	$request->content_type('application/json');
	$request->content( $json_in );
	
	my $user_agent = LWP::UserAgent->new();
	my $response = $user_agent->request($request);
	
	if ( !$response->is_success()
	  && !$args{'content_on_error'} # some calls ( eg Report.Get ) now return an error when there is no error
	) {
        my $croak_text = $response->status_line();
		if ( $response->content ) {
            $croak_text .= "\n" . $response->content;
        }
		croak "Request failed:" . $croak_text
    }
    
	carp "Response >" . ( defined( $response ) ? $response->content() : '' ) . "<"
		if $verbose;

	my $json_out;
	
	if ( exists( $METHODS_RETURNING_INVALID_JSON{ $args{'method'} } ) )
	{
		$json_out = $response->content();
	}
	else
	{
		$json_out = JSON::decode_json( $response->content() );
	}
	
	carp "JSON Response >" . ( defined( $json_out ) ? Dumper($json_out) : '' ) . "<"
		if $verbose;
	
	if ( $test_mode )
	{
		$json_out = undef;
	}
	
	return $json_out;
}


=head2 verbose()

Control the verbosity of the debugging output.

$site_catalyst->verbose( 1 ); # turn on verbose information

$site_catalyst->verbose( 0 ); # quiet!

warn 'Verbose' if $site_catalyst->verbose(); # getter-style

=cut

sub verbose
{
	my ( $self, $verbose ) = @_;
	
	$self->{'verbose'} = ( $verbose || 0 )
		if defined( $verbose );
	
	return $self->{'verbose'};
}


=head1 RUNNING TESTS

By default, only basic tests that do not require a connection to Adobe
SiteCatalyst's platform are run in t/.

To run the developer tests, you will need to do the following:

=over 4

=item *

Request access to Adobe web services from your Adobe Online Marketing Suite administrator.

=item *

In Adobe SiteCatalyst's interface, you will need to log in as an admin, then go
to the "Admin" tab, "Admin Console > Company > Web Services". There you can find
your "shared secret" for your username.

=item *

Your report suite IDs can be found in Adobe SiteCatalyst's interface. Visit 
"Admin > Admin Console > Report Suites".

=back

You can now create a file named SiteCatalystConfig.pm in your own directory, with
the following content:

	package Adobe SiteCatalystConfig;
	
	sub new
	{
		return
		{
			username                => 'username',
			shared_secret           => 'shared_secret',
			report_suite_id         => 'report_suite_id',
			api_subdomain           => 'api|api2', #optional. default='api'
			verbose                 => 0, # Enable this for debugging output
		};
	}
	
	1;

You will then be able to run all the tests included in this distribution, after
adding the path to Adobe SiteCatalystConfig.pm to your library paths.


=head1 AUTHOR

Jennifer Pinkham, C<< <jpinkham at cpan.org> >>.


=head1 BUGS

Please report any bugs or feature requests at L<https://github.com/jpinkham/business-sitecatalyst/issues>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

	perldoc Business::SiteCatalyst


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Business-SiteCatalyst>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Business-SiteCatalyst>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Business-SiteCatalyst>

=item * Search CPAN

L<http://search.cpan.org/dist/Business-SiteCatalyst/>

=back


=head1 ACKNOWLEDGEMENTS
Thank you to contributor dankasak.
Special thanks for technical help and contributions from fellow CPAN author
Guillaume Aubert L<http://search.cpan.org/~aubertg/>


=head1 COPYRIGHT & LICENSE

Copyright 2017 Jennifer Pinkham.

This program is free software; you can redistribute it and/or modify it
under the terms of the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;
